From 7de87eecb3a8311197a077aae02c7199a8e2599a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Micha=C5=82=20=C5=BBygowski?= <michal.zygowski@3mdeb.com>
Date: Fri, 2 Oct 2026 12:20:20 +0200
Subject: [PATCH 8/9] feat(rdv3): add the EL3 firmware TPM

Adding NRD_EL3_FTPM=1 builds the EL3 fTPM into BL31. BL31 no longer fits
in trusted SRAM and is loaded into a Root PAS region in
NRD_CSS_CARVEOUT_RESERVED (0xF7000000, 4MB), followed by the NS CRB pages
(0xF7400000). The CRB address is also used by the EDK2 ACPI tables to
indicate the address of TPM control area to the software/OS.

Signed-off-by: Michał Żygowski <michal.zygowski@3mdeb.com>
---
 .../common/include/nrd3/nrd_css_fw_def3.h     | 16 ++++++
 .../common/include/nrd3/nrd_pas_def3.h        | 14 +++++
 .../common/include/nrd3/nrd_plat_arm_def3.h   | 55 +++++++++++++++++--
 plat/arm/board/neoverse_rd/common/nrd_plat3.c |  6 ++
 .../neoverse_rd/platform/rdv3/platform.mk     | 13 +++++
 .../platform/rdv3/rdv3_bl2_setup.c            |  4 ++
 .../platform/rdv3/rdv3_bl31_setup.c           | 10 ++++
 7 files changed, 114 insertions(+), 4 deletions(-)

diff --git a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_css_fw_def3.h b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_css_fw_def3.h
index 9c23b583e..c84436ffb 100644
--- a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_css_fw_def3.h
+++ b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_css_fw_def3.h
@@ -137,6 +137,22 @@
 			MT_MEMORY | MT_RW | MT_SECURE)
 #endif
 
+#if NRD_EL3_FTPM
+/* BL2 loads BL31 into DRAM */
+#define NRD_CSS_EL3_FTPM_BL31_MMAP					\
+		MAP_REGION_FLAT(					\
+			NRD_EL3_FTPM_BL31_BASE,				\
+			NRD_EL3_FTPM_BL31_SIZE,				\
+			MT_MEMORY | MT_RW | EL3_PAS)
+
+/* Shared with the normal world, which maps it as device memory */
+#define NRD_CSS_EL3_FTPM_CRB_MMAP					\
+		MAP_REGION_FLAT(					\
+			NRD_EL3_FTPM_CRB_BASE,				\
+			NRD_EL3_FTPM_CRB_SIZE,				\
+			MT_NON_CACHEABLE | MT_RW | MT_NS | MT_EXECUTE_NEVER)
+#endif
+
 #if RESET_TO_BL31
 /*******************************************************************************
  * BL31 specific defines.
diff --git a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_pas_def3.h b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_pas_def3.h
index e6b6d0a84..97ab8c243 100644
--- a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_pas_def3.h
+++ b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_pas_def3.h
@@ -525,6 +525,20 @@
 			GPT_GPI_SECURE)
 #endif
 
+#if NRD_EL3_FTPM
+#define NRD_PAS_EL3_FTPM_BL31						\
+		GPT_MAP_REGION_GRANULE(					\
+			NRD_EL3_FTPM_BL31_BASE,				\
+			NRD_EL3_FTPM_BL31_SIZE,				\
+			GPT_GPI_ROOT)
+
+#define NRD_PAS_EL3_FTPM_CRB						\
+		GPT_MAP_REGION_GRANULE(					\
+			NRD_EL3_FTPM_CRB_BASE,				\
+			NRD_EL3_FTPM_CRB_SIZE,				\
+			GPT_GPI_NS)
+#endif
+
 #define NRD_PAS_RMM							\
 		GPT_MAP_REGION_GRANULE(					\
 			ARM_REALM_BASE,					\
diff --git a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h
index 84dca4365..04775c894 100644
--- a/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h
+++ b/plat/arm/board/neoverse_rd/common/include/nrd3/nrd_plat_arm_def3.h
@@ -55,15 +55,28 @@
  * also mapped. In BL31, for instance, three address regions on the remote
  * chips are accessed - secure ram, css device and soc device regions.
  */
+#if NRD_EL3_FTPM
+/* BL31 image and CRB pages in DRAM need their own tables */
+# define NRD_EL3_FTPM_MMAP_ENTRIES	U(2)
+# define NRD_EL3_FTPM_XLAT_TABLES	U(4)
+#else
+# define NRD_EL3_FTPM_MMAP_ENTRIES	U(0)
+# define NRD_EL3_FTPM_XLAT_TABLES	U(0)
+#endif
+
 #if defined(IMAGE_BL31)
-#  define PLAT_ARM_MMAP_ENTRIES		(13 + ((NRD_CHIP_COUNT - 1) * 3))
-#  define MAX_XLAT_TABLES		(13 + ((NRD_CHIP_COUNT - 1) * 3))
+#  define PLAT_ARM_MMAP_ENTRIES		(13 + ((NRD_CHIP_COUNT - 1) * 3) + \
+					 NRD_EL3_FTPM_MMAP_ENTRIES)
+#  define MAX_XLAT_TABLES		(13 + ((NRD_CHIP_COUNT - 1) * 3) + \
+					 NRD_EL3_FTPM_XLAT_TABLES)
 #elif defined(IMAGE_BL32)
 # define PLAT_ARM_MMAP_ENTRIES		U(8)
 # define MAX_XLAT_TABLES		U(5)
 #elif defined(IMAGE_BL2)
-# define PLAT_ARM_MMAP_ENTRIES		(19 + (NRD_CHIP_COUNT - 1))
-# define MAX_XLAT_TABLES		(14  + ((NRD_CHIP_COUNT - 1) * 2))
+# define PLAT_ARM_MMAP_ENTRIES		(19 + (NRD_CHIP_COUNT - 1) +	\
+					 NRD_EL3_FTPM_MMAP_ENTRIES)
+# define MAX_XLAT_TABLES		(14  + ((NRD_CHIP_COUNT - 1) * 2) + \
+					 NRD_EL3_FTPM_XLAT_TABLES)
 #else
 # define PLAT_ARM_MMAP_ENTRIES		U(7)
 # define MAX_XLAT_TABLES		U(7)
@@ -612,6 +625,16 @@ typedef enum sds_region_id {
  ******************************************************************************/
 
 /* Keep BL31 below BL2 in the Trusted SRAM.*/
+#if NRD_EL3_FTPM
+#if RESET_TO_BL31
+#error "NRD_EL3_FTPM is not supported with RESET_TO_BL31"
+#endif
+#define BL31_BASE			NRD_EL3_FTPM_BL31_BASE
+#define BL31_PROGBITS_LIMIT		(NRD_EL3_FTPM_BL31_BASE +	\
+					 NRD_EL3_FTPM_BL31_SIZE)
+#define BL31_LIMIT			(NRD_EL3_FTPM_BL31_BASE +	\
+					 NRD_EL3_FTPM_BL31_SIZE)
+#else /* NRD_EL3_FTPM */
 #if RESET_TO_BL31
 #define BL31_BASE			(0x63000)
 #else
@@ -621,6 +644,7 @@ typedef enum sds_region_id {
 #endif
 #define BL31_PROGBITS_LIMIT		BL2_BASE
 #define BL31_LIMIT			(ARM_BL_RAM_BASE + ARM_BL_RAM_SIZE)
+#endif /* NRD_EL3_FTPM */
 
 /*
  * The max number of regions like RO(code), coherent and data required by
@@ -745,6 +769,29 @@ typedef enum sds_region_id {
 #define PLAT_ARM_SPMC_SIZE	(UL(1500 * 1024) + UL(0x300000) + UL(0x500000))
 #define PLAT_ARM_SPMC_BASE     (RMM_BASE - PLAT_ARM_SPMC_SIZE)
 
+/*******************************************************************************
+ * EL3 fTPM region defines.
+ ******************************************************************************/
+#if NRD_EL3_FTPM
+/*
+ * With the EL3 firmware TPM, BL31 no longer fits in trusted SRAM and is
+ * loaded into a Root PAS DRAM region inside NRD_CSS_CARVEOUT_RESERVED,
+ * followed by the (non-secure) TPM CRB interface. The first 48MB of the
+ * reserved region are left free; the CRB address is also hardcoded in the
+ * EDK2 ACPI tables (RdV3El3Ftpm.h).
+ */
+#define NRD_EL3_FTPM_BL31_BASE	(ARM_DRAM1_BASE + ARM_DRAM1_SIZE -	\
+				 NRD_CSS_DRAM1_CARVEOUT_SIZE +		\
+				 UL(0x03000000))
+#define NRD_EL3_FTPM_BL31_SIZE	UL(0x00400000)	/* 4MB */
+#define NRD_EL3_FTPM_CRB_BASE	(NRD_EL3_FTPM_BL31_BASE +		\
+				 NRD_EL3_FTPM_BL31_SIZE)
+#define NRD_EL3_FTPM_CRB_SIZE	UL(0x00002000)
+/* CRB control area at +0x40, command/response buffer in the second page */
+#define NRD_EL3_FTPM_CRB_BUF_OFFSET	UL(0x1000)
+#define NRD_EL3_FTPM_CRB_BUF_SIZE	UL(0x1000)
+#endif
+
 /*******************************************************************************
  * CPER specific defines.
  ******************************************************************************/
diff --git a/plat/arm/board/neoverse_rd/common/nrd_plat3.c b/plat/arm/board/neoverse_rd/common/nrd_plat3.c
index e5cfdefd6..66201fa9a 100644
--- a/plat/arm/board/neoverse_rd/common/nrd_plat3.c
+++ b/plat/arm/board/neoverse_rd/common/nrd_plat3.c
@@ -42,6 +42,9 @@ const mmap_region_t plat_arm_mmap[] = {
 #if SPD_spmd && SPMD_SPM_AT_SEL2
 	NRD_CSS_SPM_CORE_REGION_MMAP,
 #endif
+#if NRD_EL3_FTPM
+	NRD_CSS_EL3_FTPM_BL31_MMAP,
+#endif
 #if TRUSTED_BOARD_BOOT && !RESET_TO_BL2
 	NRD_CSS_BL1_RW_MMAP,
 #endif
@@ -102,6 +105,9 @@ const mmap_region_t plat_arm_mmap[] = {
 	NRD_CSS_MAP_BL31_DTB,
 #endif
 	NRD_CSS_NS_CMN_MMAP,
+#if NRD_EL3_FTPM
+	NRD_CSS_EL3_FTPM_CRB_MMAP,
+#endif
 	{0}
 };
 #endif /* IMAGE_BL31 */
diff --git a/plat/arm/board/neoverse_rd/platform/rdv3/platform.mk b/plat/arm/board/neoverse_rd/platform/rdv3/platform.mk
index b1d48ea8d..c98948990 100644
--- a/plat/arm/board/neoverse_rd/platform/rdv3/platform.mk
+++ b/plat/arm/board/neoverse_rd/platform/rdv3/platform.mk
@@ -49,6 +49,19 @@ override NEED_BL2			:= no
 override NEED_BL32			:= no
 endif
 
+# Firmware TPM in EL3 (PoC), reachable through a CRB interface with an SMC
+# start method. Moves BL31 to DRAM.
+NRD_EL3_FTPM				?= 0
+$(eval $(call assert_boolean,NRD_EL3_FTPM))
+$(eval $(call add_define,NRD_EL3_FTPM))
+ifeq (${NRD_EL3_FTPM},1)
+ifeq (${PLAT_RESET_TO_BL31},1)
+$(error NRD_EL3_FTPM=1 is not supported with PLAT_RESET_TO_BL31=1)
+endif
+$(eval $(call add_define_val,EL3_FTPM_SUPPORT,1))
+include drivers/tpm/ftpm/ftpm.mk
+endif
+
 # RD-V3 platform uses GIC-700 which is based on GICv4.1
 GIC_ENABLE_V4_EXTN			:= 1
 
diff --git a/plat/arm/board/neoverse_rd/platform/rdv3/rdv3_bl2_setup.c b/plat/arm/board/neoverse_rd/platform/rdv3/rdv3_bl2_setup.c
index e13ecc311..3bb56e217 100644
--- a/plat/arm/board/neoverse_rd/platform/rdv3/rdv3_bl2_setup.c
+++ b/plat/arm/board/neoverse_rd/platform/rdv3/rdv3_bl2_setup.c
@@ -61,6 +61,10 @@ static pas_region_t pas_regions[] = {
 	NRD_PAS_NS_DRAM,
 #if SPD_spmd && SPMD_SPM_AT_SEL2
 	NRD_PAS_BL32,
+#endif
+#if NRD_EL3_FTPM
+	NRD_PAS_EL3_FTPM_BL31,
+	NRD_PAS_EL3_FTPM_CRB,
 #endif
 	NRD_PAS_RMM,
 	NRD_PAS_L1GPT,
diff --git a/plat/arm/board/neoverse_rd/platform/rdv3/rdv3_bl31_setup.c b/plat/arm/board/neoverse_rd/platform/rdv3/rdv3_bl31_setup.c
index c7a505369..4355d8bf8 100644
--- a/plat/arm/board/neoverse_rd/platform/rdv3/rdv3_bl31_setup.c
+++ b/plat/arm/board/neoverse_rd/platform/rdv3/rdv3_bl31_setup.c
@@ -8,6 +8,9 @@
 #include <drivers/arm/gic600_multichip.h>
 #include <drivers/arm/rse_comms.h>
 #include <drivers/arm/smmu_v3.h>
+#if NRD_EL3_FTPM
+#include <drivers/tpm/ftpm.h>
+#endif
 
 #include <plat/arm/common/plat_arm.h>
 #include <plat/common/platform.h>
@@ -155,6 +158,13 @@ void bl31_platform_setup(void)
 	 * system control block send transations downstream to SMMUv3.
 	 */
 	initialize_plat_amu_aux_enables();
+
+#if NRD_EL3_FTPM
+	/* A failing fTPM is reported through the CRB error bit */
+	(void)ftpm_init(NRD_EL3_FTPM_CRB_BASE, NRD_EL3_FTPM_CRB_BUF_OFFSET,
+			NRD_EL3_FTPM_CRB_BUF_SIZE);
+#endif
+
 	if (smmuv3_init(NRD_CSS_GPC_SMMUV3_BASE) != 0) {
 		WARN("Failed initializing System SMMU.\n");
 	}
